tfjs-layers/src/layers/nlp/tokenizers_utils.ts

```diff
@@ -141,3 +141,162 @@ export class BytePairTokenizerCache {
+
+  return filteredTensors;
+}
+
+/**
+ * Create alternates for all special tokens that will be not split during
+ * tokenization.
+ */
+export function createAltsForUnsplittableTokens(
+  unsplittableTokens: string[]): string[] {
+
+  const prefix = 'ĵ';
+
+  // Trim out splitters.
+  const replacePattern: RegExp = /'|\s+|[^\p{L}\p{N}]+/gu;
+  return unsplittableTokens.map(
+    token => prefix + token.replace(replacePattern, ''));
+}
+
+// Typescript and TF handles special spaces differently, we need to
+// manually handle special spaces during string split.
+const SPECIAL_WHITESPACES = /\u00A0\u2009\u202f\u3000/;
+
+// String splitting regex pattern.
+const pL = 'a-zA-ZáàâäãåçéèêëíìîïñóòôöõúùûüýÿæœÁÀÂÄÃÅÇÉÈÊËÍÌÎÏÑÓÒÔÖÕÚÙÛÜÝŸÆŒĵ';
+const pN = '0-9';
+export const SPLIT_PATTERN_1 = new RegExp(
+  `'s|'t|'re|'ve|'m|'ll|'d` +
+  `|[\\s${SPECIAL_WHITESPACES.source}]+` +
+  `[\\n\\r\\t\\f६${SPECIAL_WHITESPACES.source}]| ?${pL}+|`+
+  ` ?${pN}+| ?[^\\s${pL}${pN}${SPECIAL_WHITESPACES.source}]+`,
+  'gu'
+);
+
+const SPLIT_PATTERN_2 = new RegExp(`[\\s६${SPECIAL_WHITESPACES.source}]\$`);
+
+function flatten<T>(inputs: T[][]): T[] {
+  return inputs.reduce(
+    (accumulator, value) => accumulator.concat(value), []);
+}
+
+export function regexSplit(
+  strs: string[]|string[][],
+  delimRegexPattern: RegExp | string,
+  keepDelimRegexPattern?: RegExp | string): string[][] {
+
+  if (strs[0] instanceof Array) {
+    const mapped = strs.map(arr => regexSplit(
+      arr as string[], delimRegexPattern, keepDelimRegexPattern));
+    return mapped.map((doubleArr) => flatten(doubleArr));
+  }
+
+  strs = strs as string[];
+
+  if (!(delimRegexPattern instanceof RegExp)) {
+    if (keepDelimRegexPattern) {
+      delimRegexPattern = new RegExp(
+        `(${delimRegexPattern})`, 'g');
+    }
+    return strs.map(str => str.split(delimRegexPattern).filter(s => s));
+  }
+
+  let regexPattern = delimRegexPattern as unknown as RegExp;
+  if (!regexPattern.flags.includes('g')) {
+    regexPattern = new RegExp(regexPattern.source, regexPattern.flags + 'g');
+  }
+
+  return strs.map(str => {
+    const matches = str.matchAll(regexPattern);
+
+    const splitString = [];
+    let currIdx = 0;
+    for (const match of matches) {
+      splitString.push(str.slice(currIdx, match.index));
+      if (keepDelimRegexPattern) {
+        splitString.push(
+          str.slice(match.index, match.index! + match[0].length));
+      }
+      currIdx = match.index! + match[0].length;
+    }
+    if (currIdx !== str.length) {
+      splitString.push(str.slice(currIdx, str.length));
+    }
+    return splitString.filter(s => s);
+  });
+}
+
+export async function tensorToStringArr(input: Tensor): Promise<string[]> {
+  return await input.data() as unknown as string[];
+}
+
+export async function tensorArrToString2DArr(
+  inputs: Tensor[]): Promise<string[][]> {
+  return Promise.all(inputs.map(
+    async input => await input.data() as unknown as string[]));
+}
```

Comment:
These do not appear to be used or tested. Do you plan on using them in a future PR?

Reply:
as yes, I was mainly using them in console.logs during development. In the newer branches I am working on though, however, I use them throughout the implementation (they also have since been changed to non-async functions).

I'll add some test cases in the newer PR since these function signatures are outdated now.